<!--
  This file will launch rViz along with the mmWave sensor node and configure a TI mmWave 1642 sensor using a 2D config
 -->

<launch>

  <!-- Input arguments -->
  <arg name="device" value="1642" doc="TI mmWave sensor device type [1443, 1642]"/>
  <arg name="config" value="2d" doc="TI mmWave sensor device configuration [3d_best_range_res (not supported by 1642 EVM), 2d_best_range_res]"/>
  <arg name="max_allowed_elevation_angle_deg" default="90" doc="Maximum allowed elevation angle in degrees for detected object data [0 > value >= 90]}"/>
  <arg name="max_allowed_azimuth_angle_deg" default="90" doc="Maximum allowed azimuth angle in degrees for detected object data [0 > value >= 90]}"/>

  <!-- Usb camera node -->
  <node name="usb_webcam" pkg="usb_webcam" type="usb_webcam_node" output="screen" >
    <param name="video_device" value="/dev/video0" />
    <param name="camera_info_url" value="file://$(find usb_webcam)/cfg/elp-usb8mp02g-l36.yaml" />
    <param name="image_width" value="640" />
    <param name="image_height" value="480" />
    <param name="pixel_format" value="yuyv" />
    <param name="camera_frame_id" value="usb_webcam" />
    <param name="io_method" value="mmap"/>
    <param name="framerate" value="30" />
    <param name="autoexposure" value="true" />
    <param name="autofocus" value="true" />
  </node>

  <!-- mmWave_Manager node -->
  <node pkg="node1_radarinterface" type="node1_radarinterface" name="node1_radarinterface" output="screen">
    <param name="command_port" value="/dev/ttyACM0"  />
    <param name="command_rate" value="115200"   />
    <param name="data_port" value="/dev/ttyACM1"  />
    <param name="data_rate" value="921600"   />
    <param name="max_allowed_elevation_angle_deg" value="$(arg max_allowed_elevation_angle_deg)"   />
    <param name="max_allowed_azimuth_angle_deg" value="$(arg max_allowed_azimuth_angle_deg)"   />
  </node>
  
  <!-- mmWaveQuickConfig node (terminates after configuring mmWave sensor) -->
  <node pkg="node1_radarinterface" type="mmWaveQuickConfig" name="mmWaveQuickConfig" args="$(find node1_radarinterface)/cfg/node1_radarinterface.cfg" output="screen" />

  <!-- Static transform from map to base_radar_link for visualization of stand-alone mmWave sensor using Rviz -->
  <node pkg="tf" type="static_transform_publisher" name="static_tf_map_to_base_radar_link" args="0 0 0 0 0 0 map node1_radarinterface_markers 100"/>

  <!-- Launch Rviz with pre-defined configuration to view mmWave sensor detected object data (color by intensity) -->
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find node1_radarinterface)/launch/radar_det_obj_color_by_intensity.rviz"/>

</launch>
